pkg/datasource/data.go

```diff
@@ -0,0 +1,439 @@
+
+	tags        []string
+	annotations map[string]string
+
+	payloadCount uint32
+
+	requestedFields map[string]bool
+
+	subscriptions []*subscription
+
+	requested bool
+
+	byteOrder binary.ByteOrder
+	lock      sync.RWMutex
+}
+
+func newDataSource(t Type, name string) *dataSource {
+	return &dataSource{
+		name:            name,
+		dType:           t,
+		requestedFields: make(map[string]bool),
+		fieldMap:        make(map[string]*field),
+		byteOrder:       NativeEndian,
+		tags:            make([]string, 0),
+		annotations:     map[string]string{},
+	}
+}
+
+func New(t Type, name string) DataSource {
+	ds := newDataSource(t, name)
+	ds.registerPool()
+	return ds
+}
+
+func NewFromAPI(in *api.DataSource) (DataSource, error) {
+	ds := newDataSource(Type(in.Type), in.Name)
+	for _, f := range in.Fields {
+		ds.fields = append(ds.fields, (*field)(f))
+		if !FieldFlagUnreferenced.In(f.Flags) {
+			ds.fieldMap[f.Name] = (*field)(f)
+		}
+	}
+	if in.Flags&api.DataSourceFlagsBigEndian != 0 {
+		ds.byteOrder = binary.BigEndian
+	} else {
+		ds.byteOrder = binary.LittleEndian
+	}
+	ds.registerPool()
+	// TODO: add more checks / validation
+	return ds, nil
+}
+
+func (ds *dataSource) registerPool() {
+	ds.dPool.New = func() any {
+		d := &data{
+			Payload: make([][]byte, ds.payloadCount),
+		}
+		for i := range d.Payload {
+			d.Payload[i] = make([]byte, 0)
+		}
+		return d
+	}
+}
+
+func (ds *dataSource) Name() string {
+	return ds.name
+}
+
+func (ds *dataSource) Type() Type {
+	return ds.dType
+}
+
+func (ds *dataSource) NewData() Data {
+	return ds.dPool.Get().(Data)
+}
+
+func (ds *dataSource) ByteOrder() binary.ByteOrder {
+	return ds.byteOrder
+}
+
+func resolveNames(id uint32, fields []*field, parentOffset uint32) (string, error) {
+	if id >= uint32(len(fields)) {
+		return "", errors.New("invalid parent id")
+	}
+	out := ""
+	if FieldFlagHasParent.In(fields[id].Flags) {
+		p, err := resolveNames(fields[id].Parent-parentOffset, fields, parentOffset)
+		if err != nil {
+			return "", errors.New("parent not found")
+		}
+		out = p + "."
+	}
+	out += fields[id].Name
+	return out, nil
+}
+
+// AddStaticFields adds a statically sized container for fields to the payload and returns an accessor for the
+// container; if you want to access individual fields, get them from the DataSource directly
+func (ds *dataSource) AddStaticFields(size uint32, fields []StaticField) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	idx := ds.payloadCount
+
+	// temporary write to newFields to not write to ds.fields in case of errors
+	newFields := make([]*field, 0, len(fields))
+
+	parentOffset := len(ds.fields)
+	parentFields := make(map[int]struct{})
+	checkParents := make(map[*field]struct{})
+
+	for _, f := range fields {
+		if _, ok := ds.fieldMap[f.FieldName()]; ok {
+			return nil, fmt.Errorf("field %q already exists", f.FieldName())
+		}
+		fieldName := f.FieldName()
+		nf := &field{
+			Name:         fieldName,
```

Comment:
Variable `fieldName` doesn't get used elsewhere. The variable could be placed before the `if` and be used there.

Or remove it and use `Name: f.FieldName()`

Reply:
> The variable could be placed before the if and be used there.

Looks like that was the plan 🤣